### PR TITLE
Per-API Key Caches 

### PR DESCRIPTION
**Pull request by aarsilv**
_Eppo Internal:_
🎟️ **Ticket:** [FF-2230 - JS SDK - local and chrome storage caches should be isolated per-API key](https://linear.app/eppo/issue/FF-2230/js-sdk-local-and-chrome-storage-caches-should-be-isolated-per-api-key)

During development, different environments (and thus API keys) may be used on the same device. By isolating our provided persistent storages (Local Storage, Chrome Storage, and Assignment Cache) at the per-API key level, we will prevent an SDK initialized for one environment from loading a cached configuration from a different environment.

### CODE REVIEW

File: src/index.spec.ts [hunk 1]
```diff
@@ -550,7 +550,6 @@ describe('initialization options', () => {
     let client = await init({
       apiKey,
       baseUrl,
-      assignmentLogger: mockLogger,
```
Comment (aarsilv) [inline src/index.spec.ts:553]: Didn't need this 🪓

File: src/local-storage-assignment-cache.ts [hunk 2]
```diff
@@ -2,8 +2,14 @@ import { AssignmentCache } from '@eppo/js-client-sdk-common';
 
 import { hasWindowLocalStorage } from './configuration-factory';
 
+// noinspection JSUnusedGlobalSymbols (methods are used by common repository)
```
Comment (aarsilv) [inline src/local-storage-assignment-cache.ts:5]: stop Webstorm for yelling at me for unused methods

File: src/local-storage.spec.ts [hunk 3]
```diff
@@ -15,20 +15,33 @@ describe('EppoLocalStorage', () => {
     items: ['red'],
   };
 
-  const storage = new LocalStorageBackedAsyncStore<ITestEntry>(window.localStorage);
-
   beforeEach(() => {
     window.localStorage.clear();
   });
 
-  describe('get and set', () => {
```
Comment (aarsilv) [inline src/local-storage.spec.ts:24]: opted to remove the nested `describe()`

File: src/index.ts [hunk 4]
```diff
@@ -179,9 +179,15 @@ export async function init(config: IClientConfig): Promise<IEppoClient> {
     EppoJSClient.instance.stopPolling();
     // Set up assignment logger and cache
     EppoJSClient.instance.setLogger(config.assignmentLogger);
+
+    // Note that we use the first 8 characters of the API key to create per-API key persistent storages
+    const storageKeySuffix = config.apiKey.replace(/\W/g, '').substring(0, 8);
```
Comment (leoromanovsky) [inline src/index.ts:184]: 👍

File: src/index.ts [hunk 5]
```diff
@@ -179,9 +179,15 @@ export async function init(config: IClientConfig): Promise<IEppoClient> {
...
+
     // default behavior is to use a LocalStorage-based assignment cache.
     // this can be overridden after initialization.
-    EppoJSClient.instance.useCustomAssignmentCache(new LocalStorageAssignmentCache());
+    EppoJSClient.instance.useCustomAssignmentCache(
+      new LocalStorageAssignmentCache(storageKeySuffix),
+    );
```
Comment (leoromanovsky) [inline src/index.ts:190]: FYI - a factory is needed here as well https://linear.app/eppo/issue/FF-2113/create-chromestorage-assignment-cache

File: src/local-storage-assignment-cache.spec.ts [hunk 6]
```diff
@@ -49,4 +49,69 @@ describe('LocalStorageAssignmentCache', () => {
       }),
     ).toEqual(false); // this key has not been logged
   });
+
+  it('can have independent caches', () => {
```
Comment (leoromanovsky) [inline src/local-storage-assignment-cache.spec.ts:53]: 👨‍🍳

File: src/local-storage-assignment-cache.ts [hunk 7]
```diff
@@ -2,8 +2,14 @@ import { AssignmentCache } from '@eppo/js-client-sdk-common';
...
 class LocalStorageAssignmentShim {
-  LOCAL_STORAGE_KEY = 'EPPO_LOCAL_STORAGE_ASSIGNMENT_CACHE';
+  private readonly localStorageKey: string;
+
+  public constructor(storageKeySuffix?: string) {
```
Comment (leoromanovsky) [inline src/local-storage-assignment-cache.ts:9]: did you consider making this a required parameter?

**Review by leoromanovsky — APPROVED** [review]

**Review by greghuels — APPROVED** [review]
I love the tests. Approved, with the caveat that I'm also looking at this repo for the first time.

File: src/local-storage-assignment-cache.ts (hunk 7 above)
Comment (aarsilv) [reply src/local-storage-assignment-cache.ts:9]: Thought about it. I suppose I should since we never really use it without one.

Comment (aarsilv) [reply src/local-storage-assignment-cache.ts:9]: Making this required caught a place I forgot to pass it through! Excellent suggestion!

**Pull request merged**